### PR TITLE
Save mutated arguments after the kernel call in generated autograd kernels

### LINKED ISSUE #193671: Silent wrong gradients: nn.RReLU under non-reentrant checkpointing saves uninitialized noise
### 🐛 Describe the bug

`F.rrelu` / `nn.RReLU` with `training=True` produces **silently wrong gradients** under
`torch.utils.checkpoint.checkpoint(..., use_reentrant=False)`, and under any
`saved_tensors_hooks` that materializes the tensor at pack time (e.g. `save_on_cpu`).
The forward output is bit-exact, so nothing looks wrong until gradients are compared.

```python
import contextlib
import torch
import torch.nn.functional as F
from torch.autograd.graph import saved_tensors_hooks
from torch.utils.checkpoint import checkpoint, set_checkpoint_early_stop

def rrelu(t):
    return F.rrelu(t, 0.1, 0.9, training=True)

def grad(outer=contextlib.nullcontext(), use_checkpoint=False):
    torch.manual_seed(0)
    x = torch.randn(4096, requires_grad=True)
    with outer:
        y = checkpoint(rrelu, x, use_reentrant=False) if use_checkpoint else rrelu(x)
        y.sum().backward()
    return x.grad

ref = grad()
clone_hooks = saved_tensors_hooks(lambda t: t.clone(), lambda t: t)

print("checkpoint    :", (grad(use_checkpoint=True) - ref).abs().max().item())
print("saved hooks   :", (grad(outer=clone_hooks) - ref).abs().max().item())
print("no early stop :", (grad(set_checkpoint_early_stop(False), True) - ref).abs().max().item())

# The forward is unaffected, which is why this is silent.
torch.manual_seed(0)
a = torch.randn(4096, requires_grad=True)
out_ckpt = checkpoint(rrelu, a, use_reentrant=False)
torch.manual_seed(0)
b = torch.randn(4096, requires_grad=True)
out_eager = rrelu(b)
print("forward diff  :", (out_ckpt - out_eager).abs().max().item())
```

Observed on `2.15.0.dev20260815+cpu`:

```
torch 2.15.0.dev20260815+cpu
checkpoint    : 0.8278336524963379
saved hooks   : 1.0
no early stop : 0.0
forward diff  : 0.0
```

Expected: the first three are `0.0`. With the default `preserve_rng_state=True` the
recomputed noise is identical to the original forward, so the gradients must match exactly.
`no early stop` and `forward diff` being `0.0` is the point: the checkpointed forward is
bit-exact and the error disappears under `set_checkpoint_early_stop(False)`, so neither an
RNG probe nor an output comparison can see this.

The `checkpoint` line is stable across runs on my machine, so the snippet should reproduce
directly. The `saved hooks` line is not: five consecutive runs gave `2.46e24`, `1.0`, `1.0`,
`521.83` and `8336.83`, which is what reading uninitialized memory looks like.

Inspecting the gradient rather than its error norm shows the dangerous failure mode. In one
`saved hooks` run, `x.grad.count_nonzero()` came back as 2 out of 4096: the sampled slopes
had been replaced by an almost entirely zero-filled buffer. The gradient is quietly zeroed
rather than turning into NaN or inf, so the layer just stops learning and nothing raises an
alarm.

On a CIFAR ResNet-18 with gradient checkpointing enabled I measured a relative L2 error of
0.38 on the rrelu gradients; in another isolated run individual elements reached 1e28.

### Root cause

`F.rrelu` dispatches to `aten::rrelu_with_noise(Tensor self, Tensor(b!) noise, ...)`. `noise`
is an output argument: the kernel writes the sampled negative slopes into it and
`rrelu_with_noise_backward` scales the incoming gradient with them. The generated autograd
kernel constructs `SavedVariable(noise, false)` in the `setup_derivative` block, i.e. **before**
the kernel call, so a pack hook observes the uninitialized `at::empty_like` buffer allocated in
`at::native::rrelu` (`aten/src/ATen/native/Activation.cpp`). With no hooks this is harmless,
because the `SavedVariable` simply aliases the buffer the kernel later fills.

Under non-reentrant checkpointing it is not harmless: during recomputation that same pack point
raises `_StopRecomputationError` (`torch/utils/checkpoint.py`) once the saved-tensor counter
reaches the number of tensors saved in the original forward. When `noise` is among the last
saved tensors of the segment, the replay is aborted before the kernel runs, so the buffer is
never filled.

Because `rrelu_with_noise` is `SchemaKind.mutable` rather than inplace/out, no
`increment_version(noise)` is emitted, so `SavedVariable`'s staleness check cannot catch this
either — hence the silence.

Setting `set_checkpoint_early_stop(False)` makes the error disappear, which confirms the
mechanism.

Separately, the derivative for the autogenerated `rrelu_with_noise_functional` refers to the
`noise` **input**, but that op clones `noise` and writes the slopes into the `noise_out` return,
leaving the input untouched.

I have a fix and can send a PR once this is triaged.

### Versions

Collecting environment information...
PyTorch version: 2.15.0.dev20260815+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro (10.0.19045 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.13 | packaged by Anaconda, Inc. | (main, Jul  9 2026, 14:26:47) [MSC v.1942 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2060
Nvidia driver version: 560.94
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Name: AMD Ryzen 5 5600 6-Core Processor              
Manufacturer: AuthenticAMD
Family: 107
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3501
MaxClockSpeed: 3501
L2CacheSize: 3072
L2CacheSpeed: None
Revision: 8450

Versions of relevant libraries:
[pip3] numpy==2.5.2
[pip3] torch==2.15.0.dev20260815+cpu
[conda] numpy                      2.5.2                   pypi_0           pypi
[conda] torch                      2.15.0.dev20260815+cpu  pypi_0           pypi

cc @soulitzer @ezyang @albanD @gqchen @nikitaved @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by yinjiew**
## Issue

Fixes #193671

## Summary

Root cause and discussion are in the issue. In short: `aten::rrelu_with_noise`
takes `noise` as a mutable output argument, and the generated autograd kernel
saved it before the kernel filled it. That works only while the `SavedVariable`
aliases the buffer; a hook that materializes at pack time, or a non-reentrant
checkpoint replay that early-stops at that pack point, captures uninitialized
memory instead and backward silently scales gradients by garbage. This moves the
`SavedVariable` construction for operator-mutated arguments to after the call.

Only four generated kernels change: `rrelu_with_noise`, `rrelu_with_noise_`,
`_native_batch_norm_legit` and `_batch_norm_with_update`. The two batch norm ops
read the running stats in backward only when `training=False`, when the kernel
does not update them, so they are numerically unaffected.

Also fixes the `rrelu_with_noise_functional` derivative, which referenced the
untouched `noise` input rather than the `noise_out` return.

One thing worth flagging for review: the new save site does not apply the
`guard_for` predicate that `emit_save_inputs` uses to skip saving tensors a
backward will not need, because `guard_for` is local to that closure. It is a
no-op for all four ops today (rrelu has a single arg with a derivative, so
`guard_for` bails on `len(args_with_derivatives) <= 1`; the batch norm formulas
are multi-output without `wrap_opt_if`). Happy to hoist `guard_for` to
`emit_body` scope if you would rather have it applied uniformly.

## Reproducer, before / after

Reproducer from the issue, on a CPU source build of `8f988c9c`.

Before, five consecutive runs:

    torch 2.15.0a0+git8f988c9
    checkpoint    : 0.0
    saved hooks   : nan
    no early stop : 0.0
    forward diff  : 0.0
    --- run 1 ---  checkpoint : 1.2183011048223146e+20   saved hooks : 1.2183011048223146e+20
    --- run 2 ---  checkpoint : 5.938071552281598e+34    saved hooks : 1.0
    --- run 3 ---  checkpoint : 2.457677367050863e+21    saved hooks : 2.458235813404657e+21
    --- run 4 ---  checkpoint : 6.766431746933327e+24    saved hooks : 6.767575445065897e+24
    (no early stop : 0.0 on every run)

After, same five runs:

    torch 2.15.0a0+git8f988c9
    checkpoint    : 0.0
    saved hooks   : 0.0
    no early stop : 0.0
    forward diff  : 0.0
    (all five runs identical)

## Checklist

- [x] Passes lint (`lintrunner` reports no issues on the changed files)
- [x] Added/updated tests
- [ ] Updated documentation (if applicable)
- [ ] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No.

### CODE REVIEW

**Comment by yinjiew** [comment]
@albanD  The workflow runs are still awaiting maintainer approval, so CI has not run on the current head (969ef0d). I force-pushed shortly after opening this to drop a ‘Co-Authored-By’ trailer that EasyCLA could not accept, which invalidated the earlier approval. Could you approve the runs when you get a chance? Thanks.
